### PR TITLE
feat: add command menu and keyboard shortcuts

### LINKED ISSUE #317: Keyboard shortcuts
## Feature Description

As a user, I'd like to have keyboard shortcuts to perform actions quickly.

## Use Case

Suggested shortcuts:
- Add a new document
- Open `Settings`
- View all `Documents`

## Proposed Solution

- could use [this component](https://github.com/pacocoursey/cmdk)?

## Checklist

- [x] I have searched the existing feature requests to make sure this is not a duplicate.
- [x] I have provided a detailed description of the requested feature.
- [x] I have explained the use case or scenario for this feature.
- [x] I have included any relevant technical details or design suggestions.
- [x] I understand that this is a suggestion and that there is no guarantee of implementation.

### PR DESCRIPTION
**Pull request by G3root**
fixes #317

[Screencast 2023-08-31 22:55:18.webm](https://github.com/documenso/documenso/assets/84864519/9b3799b0-fa8e-4cb0-9cf4-0e43ae2f8401)

## how to test

1. press `⌘+k` to open the command menu
2. press `N+S` simultaneously to to navigate to the settings page
3. press `N+D` simultaneously to to navigate to the documents page

## Summary by CodeRabbit

New Features:
- Added a new `CommandMenu` component to the dashboard, providing users with a quick and easy way to navigate to different pages and change the theme.
- Enhanced the `CommandDialog` component with a new `commandProps` prop, allowing for more customization options.

Refactor:
- Updated the dashboard layout to include the `CommandMenu` component, improving the user interface and overall user experience.

### CODE REVIEW

File: apps/web/package.json [hunk 1]
```diff
@@ -36,6 +36,7 @@
     "react-dom": "18.2.0",
     "react-dropzone": "^14.2.3",
     "react-hook-form": "^7.43.9",
+    "react-hotkeys-hook": "^4.4.1",
```
Comment (G3root) [inline apps/web/package.json:39]: hook to add hotkeys

File: packages/ui/primitives/command.tsx [hunk 2]
```diff
@@ -25,13 +25,18 @@ const Command = React.forwardRef<
...
+  commandProps?: React.ComponentPropsWithoutRef<typeof CommandPrimitive>;
+};
 
-const CommandDialog = ({ children, ...props }: CommandDialogProps) => {
+const CommandDialog = ({ children, commandProps, ...props }: CommandDialogProps) => {
   return (
     <Dialog {...props}>
       <DialogContent className="overflow-hidden p-0 shadow-2xl">
-        <Command className="[&_[cmdk-group-heading]]:text-muted-foreground [&_[cmdk-group-heading]]:px-2 [&_[cmdk-group-heading]]:font-medium [&_[cmdk-group]:not([hidden])_~[cmdk-group]]:pt-0 [&_[cmdk-group]]:px-2 [&_[cmdk-input-wrapper]_svg]:h-5 [&_[cmdk-input-wrapper]_svg]:w-5 [&_[cmdk-input]]:h-12 [&_[cmdk-item]]:px-2 [&_[cmdk-item]]:py-3 [&_[cmdk-item]_svg]:h-5 [&_[cmdk-item]_svg]:w-5">
+        <Command
```
Comment (G3root) [inline packages/ui/primitives/command.tsx:36]: add the ability to add custom props for the command component and reduce height and width size from 5 to 4 for `[cmdk-item]_svg`

**Comment by Mythie** [comment]
@catalinpit can you run with this and. bring it up to speed?

Ideally we avoid the `preventDefaults` if possible, using the command bar design that is currently in the design phase for a first cut.

File: apps/web/src/components/(dashboard)/common/command-menu.tsx [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+  CommandShortcut,
+} from '@documenso/ui/primitives/command';
+
+export function CommandMenu() {
+  const { setTheme } = useTheme();
+  const { push } = useRouter();
+  const [open, setOpen] = useState(false);
+  const [search, setSearch] = useState('');
+  const [pages, setPages] = useState<string[]>([]);
+  const page = pages[pages.length - 1];
```
Comment (Mythie) [inline apps/web/src/components/(dashboard)/common/command-menu.tsx:27]: `currentPage`

File: apps/web/src/components/(dashboard)/common/command-menu.tsx [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+
+  useHotkeys('ctrl+k', (e) => {
+    e.preventDefault();
```
Comment (Mythie) [inline apps/web/src/components/(dashboard)/common/command-menu.tsx:30]: Do we need to `preventDefault`?

File: apps/web/src/components/(dashboard)/common/command-menu.tsx [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+    setOpen((open) => !open);
+  });
+  useHotkeys('n+s', () => push('/settings'));
+  useHotkeys('n+d', () => push('/documents?status=ALL'));
```
Comment (Mythie) [inline apps/web/src/components/(dashboard)/common/command-menu.tsx:34]: Should consider making these constants at the top of the file, not that it'll do anything but it'll look more deliberate.

File: apps/web/src/components/(dashboard)/common/command-menu.tsx [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+
+  return (
+    <CommandDialog
+      commandProps={{
+        onKeyDown: (e) => {
+          // Escape goes to previous page
+          // Backspace goes to previous page when search is empty
+          if (e.key === 'Escape' || (e.key === 'Backspace' && !search)) {
+            e.preventDefault();
```
Comment (Mythie) [inline apps/web/src/components/(dashboard)/common/command-menu.tsx:43]: This one makes sense, what happens if we're at the last page though? Shouldn't we close the dialog 🤔

File: apps/web/src/components/(dashboard)/common/command-menu.tsx [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+      onOpenChange={setOpen}
+    >
+      <CommandInput
+        value={search}
+        onValueChange={setSearch}
+        placeholder="Type a command or search..."
+      />
+      <CommandList>
+        <CommandEmpty>No results found.</CommandEmpty>
+        {!page && (
```
Comment (Mythie) [inline apps/web/src/components/(dashboard)/common/command-menu.tsx:58]: This is pretty cool, we should be mindful to split these out in the future as we add more.

File: apps/web/src/components/(dashboard)/common/command-menu.tsx [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+  useHotkeys('ctrl+k', (e) => {
+    e.preventDefault();
+    setOpen((open) => !open);
+  });
+  useHotkeys('n+s', () => push('/settings'));
+  useHotkeys('n+d', () => push('/documents?status=ALL'));
+
+  return (
+    <CommandDialog
+      commandProps={{
```
Comment (Mythie) [inline apps/web/src/components/(dashboard)/common/command-menu.tsx:38]: We could probably do something with typescript here to accept the subset of `commandbar` props we're interested in but I'm happy to stick with this for now.

**Pull request merged**